### PR TITLE
add babel-plugin-transform-annotation-to-extern

### PR DESCRIPTION
**Pull request by erwinmombay**
initial PR for typedef extraction. Partial work for https://github.com/ampproject/amphtml/issues/23425

This PR locates typedefs and constructs the code for the extern.

Next Steps: Put back type information from parsed doctrine AST into constructor declaration and field declaration.

NOTE: This transformer should actually not do any transformation to source and only accumulate the code for the extern.

### CODE REVIEW

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        )[0].type;
+
+        // We can assume theres only 1 variable declarator
+        // if we can find a typedef leading comment.
+        const id = node.declarations[0].id.name;
+        const funcDecl = buildFuncDecl(t, id);
+        const memberExprs = fields.map(field => {
+          return buildExprStmt(t, id, field.key);
+        });
+        path.replaceWithMultiple([funcDecl, ...memberExprs]);
```
Comment (erwinmombay) [inline build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: Im only doing this for test purposes.

@jridgewell is there a way to emit the source code without modifying the current Program?

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/test/fixtures/extract-annotations/transform-typedef/output.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ *
+ * @typedef {{
+ *   viewportRect: !LayoutRectDef,
+ *   targetRect: !LayoutRectDef,
+ * }}
+ */
+function PositionEntryDef() {}
+
+PositionEntryDef.prototype.viewportRect;
```
Comment (jridgewell) [inline build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/test/fixtures/extract-annotations/transform-typedef/output.js]: Why are these being converted to prototype properties?

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+const doctrine = require('@jridgewell/doctrine');
+
+const flatten = arr => [].concat(...arr);
+
+const buildFuncDecl = (t, id) => {
+  return t.functionDeclaration(t.identifier(id), [], t.blockStatement([]));
+};
+
+const buildExprStmt = (t, id, field) => {
+  return t.expressionStatement(
```
Comment (jridgewell) [inline build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: You can use `babel.template` to make these ast tree much more naturally:

```js
const {template} = babel;

const buildExprStmt = template`BASE.prototype.PROP;`;

const ast = buildExprStmt({
  BASE: id,
  PROP: field,
});

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  const {types: t} = babel;
+  return {
+    visitor: {
+      VariableDeclaration(path) {
+        const {node} = path;
+        if (!node.leadingComments) {
+          return;
+        }
+
+        const comments = node.leadingComments.filter(comment => {
```
Comment (jridgewell) [inline build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: Use `find` instead of `filter`.

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js (hunk 1 above)
Comment (jridgewell) [reply build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: I don't understand? What are you trying to do?

**Comment by erwinmombay** [comment]
@jridgewell im trying to take the typedefs in certain files and generate an extern instead which will force closure to not obfuscate the fields. 

this would stop something like the following object structure and its fields from getting optimized during a cast:

```
/** @type {MyTypeDef} */ (someJsonData)
```

is there a different way for them to be declared in an extern? im mostly following what we have in https://github.com/ampproject/amphtml/blob/master/build-system/amp.extern.js

Im not trying to do an actual transformation on the source (besides typedef removal), and ill be caching the AST (or the output source generated) and make a temp file to be piped into closure.

**Comment by jridgewell** [comment]
You'll have to just change the comment, then?

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/test/fixtures/extract-annotations/remove-typedefs/options.json [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{
+  "plugins": ["../../../../../babel-plugin-transform-annotation-to-extern"]
```
Comment (jridgewell) [inline build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/test/fixtures/extract-annotations/remove-typedefs/options.json]: Need `emitTypedefs`?

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+const buildTypedefs = (t, path) => {
+  const typedefs = [];
+  for (const [typedefName, typedefComment] of TYPEDEFS) {
+    const ast = buildVarDeclAndComment(t, path, typedefName, typedefComment);
+    typedefs.push(ast);
+  }
+  return typedefs;
+};
+
+const buildVarDeclAndComment = (t, path, name, comment) => {
```
Comment (jridgewell) [inline build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: `path` is unused.

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    visitor: {
+      Program: {
+        exit(path) {
+          if (shouldWriteToFile) {
+            // Stub
+            console.log('write to temp directory');
+          }
+          if (shouldEmitTypedefs) {
+            // Preserve the leading LICENSE comment.
+            path.addComment('leading', path.parent.comments[0].value);
```
Comment (jridgewell) [inline build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: This is only for tests, why add a trailing comment and leave the program intact?

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+          return;
+        }
+
+        // We can assume theres only 1 variable declaration  when a typedef
+        // annotation is found. This is because Closure Compiler does not allow
+        // declaration of multiple variables with a shared type information.
+        const typedefName = node.declarations[0].id.name;
+
+        const typedefLocation = TYPEDEFS.get(typedefName);
+        if (!typedefLocation) {
```
Comment (jridgewell) [inline build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: `has`?

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+          TYPEDEFS.set(typedefName, typedefComment.value);
+        }
+
+        // We can't easily remove comment nodes so we just empty the string out.
+        typedefComment.value = '';
```
Comment (jridgewell) [inline build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: No need to do this if we're removing the path.

Comment (erwinmombay) [reply build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: doesn't seem to remove the leading comments if i don't empty it out:

see https://astexplorer.net/#/gist/d501eb1193748a09482f7cdd69e1870e/ba9a293cedd23f9c8b64bd16d114e0cee2d503cf

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js (hunk 7 above)
Comment (erwinmombay) [reply build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: do you mean to not replace the program or to just not add the comment? if the latter its mostly to appease the linter

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js (hunk 1 above)
Comment (erwinmombay) [reply build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: we've discussed that typedefs can live in externs and i've fixed the code.

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js (hunk 7 above)
Comment (jridgewell) [reply build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: Don't do anything to the program, just postpend the typedefs:

```js
path.addComment('trailing', typdefsComment)
```

Comment (erwinmombay) [reply build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: done.

**Comment by erwinmombay** [comment]
@jridgewell applied recommendations

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js [hunk 10]
```diff
@@ -93,12 +95,13 @@ module.exports = function(babel) {
 
         if (!TYPEDEFS.has(typedefName)) {
           TYPEDEFS.set(typedefName, typedefComment.value);
+          // We can't easily remove comment nodes so we just empty the string
+          // out.
+          typedefComment.value = '';
```
Comment (jridgewell) [inline build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: This still shouldn't be necessary if we're removing the path.

File: build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js [hunk 11]
```diff
@@ -93,12 +95,13 @@ module.exports = function(babel) {
 
         if (!TYPEDEFS.has(typedefName)) {
```
Comment (jridgewell) [inline build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: Isn't this an error case?

Comment (erwinmombay) [reply build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: @jridgewell do you mean where theres a duplicate typedef name?

Comment (jridgewell) [reply build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: Yes

Comment (erwinmombay) [reply build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: @jridgewell it should but the linter should catch it from a project level, and in program level it should be an error too.

I'm probably clearing the Map in the wrong place, im currently doing it in `pre` (is that executed only once?), should i be doing it on Program enter?

Comment (jridgewell) [reply build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: I would throw here.

Pre is executed once per plugin. `Program:enter` is executed once per file.

Comment (erwinmombay) [reply build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js]: done.

**Comment by erwinmombay** [comment]
@jridgewell PTAL

File: build-system/babel-plugins/babel-plugin-transform-html-template/index.js [hunk 12]
```diff
@@ -65,9 +65,6 @@ module.exports = function({types: t}) {
   return {
     name: 'transform-html-templates',
     visitor: {
-      Program() {
-        INSERTED_TEMPLATES.clear();
```
Comment (jridgewell) [inline build-system/babel-plugins/babel-plugin-transform-html-template/index.js]: Why was this removed?

Comment (erwinmombay) [reply build-system/babel-plugins/babel-plugin-transform-html-template/index.js]: wasnt intentional. reverting change.
